### DISCUSSION TITLE
Can people be bodies?

### DISCUSSION THREAD

**Discussion by johncrater** [dyn4j/dyn4j#299] [Q&A]
I'm writing a football game and I want to use Dyn4j for both the physics engine and for collision detection. There are some differences that I see with this from the examples and I'd like some advice. First, people have their own internal direction in space that is more than just response to external forces. So, I have found that if I leave too much up to the physics engine, my players start to behave like billiard balls. For example, I had a player run out to the team huddle and he smashed into his team mates sending them rolling across the field. But, if I try to control more of their movements outside the engine, the engine starts to seem less useful. Another example, is I had a player stuck behind his teammates and unable to move through them to get on to the field for a play. 

I'm sure you've had experience with this kind of application of dyn4j. Is it an appropriate tool for this kind of job? Should I try another approach? Any advice?

**Comment by wnbittle** [comment]
From your description, I'm assuming a top down type game.  For something like this, you'll need to tweak things a bit and you may have to do things that won't make physical sense.  A few questions:

- Is the game will be multi-player or single-player?
- Are the people directly controlled?  Can the player move a person directly or only give orders like "go to the huddle"?
- What physical interactions _are_ you looking for?

If you don't need physical interaction for most of the bodies in the game, you can make them all "sensors".  This is a property on the `Fixture`/`BodyFixture` classes that will still track collisions, but not resolve them.  You can then manually resolve them however you'd like.  If you don't need physical resolution at all, you can extends the `AbstractCollisionWorld` class and implement the one method there and handle the collisions yourself.

Player control and realism are actually at odds with each other and you have to tweak things heavily sometimes to get what you want.  For example, one thing I would suggest is to add to every `Body` a large amount of linear damping to simulate ground friction - almost to the point you can barely move them.  Then, when a player wants to move a person, you use the `MotorJoint` with equally high maxForce (and other properties) to make the person move quickly, but it should stop quickly as well.  This should also force the other people to stop quickly if they are impacted.

If the people aren't directly player controlled and instead are controlled by commands like "go here", you should implement a motion control that includes an easing function.  Going back to realism for a moment, a person on the field doesn't immediately stop or start moving at their fastest pace, they ramp up or slow down.  Easing functions would allow a smooth transition to stopped (for the huddle) so it doesn't overshoot and bowl into the other people.

I'm not sure why one person would get stuck behind others, but one thought is to make the player's collision shape smaller than their visible image.  This can help with reducing getting stuck and also adds a little more realism (people on the field don't purposefully ram their shoulders into the other people trying to get by).

Another thought is to toggle properties like the `MassType` and `Mass` when people have finished their movement.  For example, imagine you are sending 8 people to the huddle on the field.  When any person makes it to their huddle position, set their `MassType` to `INFINITE`.  Another example, when a player is trying to enter the field, but other players are in their way, if those other players are not part of the play, set their `mass` to something a little smaller than the player entering the field so that they move out of the way.

It's a little hard to provide advise with the detail you've provided, but hopefully these give you some ideas or things to try.

**Comment by johncrater** [comment]
Hi William. Thanks for the great and rapid answer. To answer your
questions, the game is top down, single user and the user selects plays but
does direct them real time. The players of the field then do their thing
without user intervention. I already have algorithms for accelerating and
decelerating realistically which I put in before investigating  dyn4j. The
complexity of implementing these algorithms was partly responsible for me
seeking out something like dyn4j to begin with.

In my game loop, I calculate acceleration and deceleration and then just
set the linear velocity of the players/bodies in the game prior to updating
the engine. This seems to work pretty well although, like I say, I am
trying to transition away from implementing the physics myself as I learn
more about dyn4j. Changing the mass for players at rest sounds like a good
idea. Right now, I just have them setEnabled(false) which makes them like
ghosts so players can move through them to get to the field. Your other
suggestions seem promising as well. .

I am just getting into collision detection. What is the difference between
a collision and contact?

Thanks for the help and great job with the library. I really enjoy using
it.

**Reply by wnbittle** [reply]
A collision is two bodies or fixtures overlapping.  A collision can have one or two contact points.  Contacts are what get "solved" in the physics pipeline

**Comment by JollyNabeel** [comment]
Great question! For a deeper look at the relationship between identity and existence, check out [FC Mobile Latest Version Download](https://fcmobileapkk.com/)
